src/ptvsd/_vendored/pydevd/pydevd.py

```diff
@@ -546,7 +546,31 @@ def handle_breakpoint_expression(self, pybreakpoint, info, new_frame):
             if val is not None:
                 info.pydev_message = str(val)
 
-    def get_file_type(self, abs_real_path_and_basename):
+    def _internal_get_file_type(self, abs_real_path_and_basename):
+        basename = abs_real_path_and_basename[-1]
+        if basename.startswith('<frozen '):
+            # In Python 3.7 "<frozen ..." appear multiple times during import and should be
+            # ignored for the user.
+            return self.PYDEV_FILE
+        return self._dont_trace_get_file_type(basename)
+
+    def dont_trace_external_files(self, abs_path):
+        '''
+        :param abs_path:
+            The result from get_abs_path_real_path_and_base_from_file or
+            get_abs_path_real_path_and_base_from_frame.
+
+        :return
+            True :
+                If files should NOT be traced.
+
+            False:
+                If files should be traced.
+        '''
+        # By default all external files are traced.
+        return False
+
+    def get_file_type(self, abs_real_path_and_basename, _cache_file_type={}):
```

Comment:
The `_cache_file_type` should to be defined in the module level (you can still assign to the local in the function definition: `_cache_file_type=_CACHE_FILE_TYPE`) and we should check that it's empty when `on_setdebuggerproperty_request` is called as the result of `get_file_type` must always be the same (there are other caches throughout the debugger which rely on having always the same file type, so, `on_setdebuggerproperty_request` cannot change that invariant -- after we started debugging the `on_setdebuggerproperty_request` for that parameter can no longer be changed).